source/common/filesystem/posix/filesystem_impl.cc

```diff
@@ -158,14 +158,17 @@ bool InstanceImplPosix::illegalPath(const std::string& path) {
     return true;
   }
 
-  // Platform specific path sanity; we provide a convenience to avoid Envoy
-  // instances poking in bad places. We may have to consider conditioning on
-  // platform in the future, growing these or relaxing some constraints (e.g.
-  // there are valid reasons to go via /proc for file paths).
-  // TODO(htuch): Optimize this as a hash lookup if we grow any further.
-  if (absl::StartsWith(canonical_path.rc_, "/dev") ||
-      absl::StartsWith(canonical_path.rc_, "/sys") ||
-      absl::StartsWith(canonical_path.rc_, "/proc")) {
+  return protectedPath(canonical_path.rc_);
+}
+
+bool InstanceImplPosix::protectedPath(const std::string& path) {
+ // Platform specific path sanity; we provide a convenience to avoid Envoy
+ // instances poking in bad places. We may have to consider conditioning on
+ // platform in the future, growing these or relaxing some constraints (e.g.
+ // there are valid reasons to go via /proc for file paths).
+ // TODO(htuch): Optimize this as a hash lookup if we grow any further.
+ if (absl::StartsWith(path, "/dev/") || path == "/dev" || absl::StartsWith(path, "/sys/") ||
+      path == "/sys" || absl::StartsWith(path, "/proc/") || path == "/proc") {
```

Comment:
I think we at the very least need a comment explaining why we make this check and might also be nice to extract the individual checks into a function, e.g. a `checkPathPrefix` function. This is what I originally thought about when I suggested adding a function, but what you've done also seems like an improvement :) 

I wonder if it might also be sufficient to just check for a `/dev/` prefix, as I'm not sure if we're worried about targeting the directories themselves?

cc @htuch since I see you in the TODO

Reply:
At least adding the `protectedPath` function allowed to do tests for things like /system as the actual `illegalPath` also checks for the existence of the target which made it a bit problematic.

I considered adding a more detailed comment however the original comment for this already explains why it is done so I moved the comment to the `protectedPath` function.

The reason I included checks for `/dev` etc. was because the original tests checked that as well so they would've failed with just `absl::StartsWith(path, "/dev/")`. I'm not really that proficient with C++ or the project but from the looks of it, `illegalPath` is called on reading some configurations mostly so it probably won't be a significant performance hit to check them both.

Reply:
I think `/dev/` prefix check is fine. We actually added these based on fuzzers that would go walking everywhere across the filesystem; it didn't seem a terribly bad idea to provide some defense against malicious configuration, e.g. trying to `DataSource` from `/dev/mem`. If someone really wanted to access files in these paths, I think we could relax.

Reply:
So since the tests seem to pass (other than the macos timeout) and this would solve my issue and doesn't seem to break existing stuff. Do you want me to do something about this?

At least removing the exact "/dev" etc. match would require adjusting the existing tests as well.

Reply:
Just switch to `/foo/` everywhere and it should be fine.

Reply:
Just having the `absl::StartsWith(path, "/foo/")` leads to another issue.

The `illegalPath` uses `canonicalPath` check for path existence before determining if the path even exists prior to checking if it's illegal or not. `canonicalPath` uses `realpath` which turns `/foo/` into `/foo` so the tests fail due to that.

This lead me to looking at the `/dev/fd/` check which in turn allows arbitrary file access without canonical path check as you can do something like `/dev/fd/../../../dev/mem` at least on Linux.

So I think either the platform specific checks that are now avoided should either be introduced or this check should be relaxed overall.

Reply:
OK, so having both exact match and prefix with `/dev/` should work.

Reply:
This still leaves the issue with /dev/fd/ check (which was not really part of the original pull request) as you can use it to circumvent any of these checks as it's not done against canonical path.

As for the check for `/dev` and prefix `/dev/` is in the second commit and it passed tests.

https://github.com/envoyproxy/envoy/pull/15878/commits/e9475af34e38bfd3500ca26433536b7a2b646262

Reply:
Maybe we should just get rid of these checks. @asraa if we drop these, the fuzzers are going to go mad unless we have a way of being to ensure any time we see a path, it gets sanitized. Do you have any thoughts on how we can generically handle this? Can we run the fuzzers in some hermetic environment where they never see the real filesystem?